### PR TITLE
fix: Fix range out of index error with a temporary workaround

### LINKED ISSUE #540: range end index 294912 out of range for slice of length 147456
### Describe the bug

I just saw this when running benchmarks with latest from main and with xxhash64 disabled.

```
Caused by: org.apache.comet.CometNativeException: General execution error with reason org.apache.comet.CometNativeException: range end index 294912 out of range for slice of length 147456
        at comet::errors::init::{{closure}}(__internal__:0)
        at std::panicking::rust_panic_with_hook(__internal__:0)
        at std::panicking::begin_panic_handler::{{closure}}(__internal__:0)
        at std::sys_common::backtrace::__rust_end_short_backtrace(__internal__:0)
        at rust_begin_unwind(__internal__:0)
        at core::panicking::panic_fmt(__internal__:0)
        at core::slice::index::slice_end_index_len_fail(__internal__:0)
        at arrow_data::transform::variable_size::build_extend::{{closure}}(__internal__:0)
        at arrow_data::transform::MutableArrayData::extend(__internal__:0)
        at comet::execution::operators::copy_array(__internal__:0)
        at comet::execution::operators::copy_or_cast_array(__internal__:0)
        at <comet::execution::operators::copy::CopyStream as futures_core::stream::Stream>::poll_next(__internal__:0)
        at <datafusion_physical_plan::stream::RecordBatchStreamAdapter<S> as futures_core::stream::Stream>::poll_next(__internal__:0)
        at Java_org_apache_comet_Native_executePlan(__internal__:0)
        at <unknown>(__internal__:0).
	at org.apache.comet.Native.executePlan(Native Method)
	at org.apache.comet.CometExecIterator.getNextBatch(CometExecIterator.scala:102)
	at org.apache.comet.CometExecIterator.hasNext(CometExecIterator.scala:125)
	at org.apache.spark.sql.comet.execution.shuffle.CometShuffleWriteProcessor.write(CometShuffleExchangeExec.scala:492)
```

### Steps to reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #540.

### CODE REVIEW

**Comment by viirya** [comment]
I will add a custom Java importer implementation to use the fixed `offset` field from arrow-rs.

File: common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  private ArrowBuf importFixedBytes(ArrowType type, int index, long bytesPerSlot) {
+    final long capacity = bytesPerSlot * fieldNode.getLength();
+    return importBuffer(type, index, capacity);
+  }
+
+  private ArrowBuf importOffsets(ArrowType type, long bytesPerSlot) {
+    final long capacity = bytesPerSlot * (fieldNode.getLength() + 1);
+    final long offset = snapshot.offset * bytesPerSlot;
+    return importBuffer(type, 1, offset, capacity);
+  }
```
Comment (viirya) [inline common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:137]: This class adds offset to buffer pointer. This is major change to Arrow Java `BufferImportTypeVisitor`.

**Comment by viirya** [comment]
Looks like I need to copy more Arrow Java classes to make it work...

**Comment by viirya** [comment]
> Looks like I need to copy more Arrow Java classes to make it work...

Instead of a real fix which requires adding `offset` support to Arrow Java, I added a hacky workaround in my arrow-rs repo, and added custom Java importer implementation.

I re-ran the query and verified this workaround fixes the range out of index error.

File: common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+          "Offset buffer for type %s is malformed: start: %s, end: %s",
+          type,
+          start,
+          end);
+      // HACK: For the issue https://github.com/apache/datafusion-comet/issues/540
+      // As Arrow Java doesn't support `offset` in C Data interface, we cannot correctly import
+      // a slice of string from arrow-rs to Java Arrow and then export it to arrow-rs again.
+      // So we add this hack to always take full length of data buffer by assuming the first offset
+      // is always 0 which is true for Arrow Java and arrow-rs.
+      final int len = end;
```
Comment (viirya) [inline common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:245]: This is the workaround added in Arrow Java side.

Comment (viirya) [reply common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:245]: A similar workaround is also added in the custom arrow-rs branch.

**Comment by viirya** [comment]
cc @andygrove This uses a temporary workaround to fix the issue. I'm not sure how long it will take to have the Arrow Java fix, so I think a workaround but not a real fix might be good for us to move forward.

File: common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java (hunk 2 above)
Comment (viirya) [reply common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:245]: This workaround assumes that the first element in offset buffer is always 0. This is true for arrow-rs and Arrow Java although Arrow spec just recommends it but not enforces it.

Comment (andygrove) [reply common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:245]: Are the changes in the custom arrow-rs branch different to the fix that went into arrow-rs master branch recently?

**Comment by andygrove** [comment]
> > Looks like I need to copy more Arrow Java classes to make it work...
> 
> Instead of a real fix which requires adding `offset` support to Arrow Java, I added a hacky workaround in my arrow-rs repo, and added custom Java importer implementation.
> 
> I re-ran the query and verified this workaround fixes the range out of index error.

I also ran the benchmarks and can confirm that this fixes the issue.

File: common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java (hunk 2 above)
Comment (viirya) [reply common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java:245]: The fix into arrow-rs repo is real fix that correctly reflecting offset of offset buffer into `offset` field in C Data interface. But only the fix is not enough, we still need the `offset` support in Arrow Java.

So the workaround here is different but just hacky solution. Instead filing correct `offset` in C Data interface, we calculate correct data buffer length in a hacky way by assuming the first offset value in offset buffer is always 0. So we get correct data buffer length even offset buffer is a slice (i.e., `offset-ed` in arrow-rs).

**Comment by viirya** [comment]
I proposed the workaround fix to arrow-rs at https://github.com/apache/arrow-rs/pull/5958.

**Comment by viirya** [comment]
@andygrove Could you verify again if the latest change still resolves the issue? I changed it to use the commit proposed for arrow-rs https://github.com/apache/arrow-rs/pull/5958 behind a feature flag.

**Comment by viirya** [comment]
I changed the arrow-rs repo to use the change at https://github.com/apache/arrow-rs/pull/5964. See if everything in CI can pass.

**Comment by andygrove** [comment]
@viirya I confirmed that this fixes the issue when I run the benchmarks locally.

**Comment by viirya** [comment]
> @viirya I confirmed that this fixes the issue when I run the benchmarks locally.

Thank you, @andygrove. Then we can merge this and wait for new arrow-rs and DataFusion releases.

**Review by andygrove — APPROVED** [review]
LGTM. Hopefully we only need to use the forks for 2-3 weeks.

**Review by andygrove — CHANGES_REQUESTED** [review]
I don't think we can depend on this external fork because the branch could get deleted. Could you create a branch in your repo instead?

File: core/Cargo.toml [hunk 3]
```diff
@@ -33,13 +33,13 @@ include = [
...
 parquet-format = "4.0.0" # This must be kept in sync with that from parquet crate
-arrow = { version = "52.0.0", features = ["prettyprint", "ffi", "chrono-tz"] }
-arrow-array = { version = "52.0.0" }
-arrow-buffer = { version = "52.0.0" }
-arrow-data = { version = "52.0.0" }
-arrow-schema = { version = "52.0.0" }
-arrow-string = { version = "52.0.0" }
-parquet = { version = "52.0.0", default-features = false, features = ["experimental"] }
+arrow = { git = "https://github.com/tustvold/arrow-rs.git", rev = "78b6139", features = ["prettyprint", "ffi", "chrono-tz"] }
+arrow-array = { git = "https://github.com/tustvold/arrow-rs.git", rev = "78b6139" }
```
Comment (andygrove) [inline core/Cargo.toml:37]: Now that the PR is merged, can we switch this to a revision of the official arrow-rs repo?

Comment (viirya) [reply core/Cargo.toml:37]: Yes. We can point to arrow-rs repo.

**Comment by viirya** [comment]
@andygrove This uses official arrow-rs repo now.

**Review by andygrove — APPROVED** [review]
Thanks @viirya

**Comment by viirya** [comment]
Thanks @andygrove

**Pull request merged**